src/core/ext/transport/binder/wire_format/wire_writer.cc

```diff
@@ -66,119 +80,301 @@ absl::Status WireWriterImpl::WriteTrailingMetadata(const Transaction& tx,
+  // because single transaction size is not required to be strictly lower than
+  // `kBlockSize`, as long as it won't overflow Android's binder buffer.
+  int64_t size = std::min<int64_t>(WireWriterImpl::kBlockSize,
+                                   data.size() - stream_tx->bytes_sent);
+  GPR_ASSERT(stream_tx->bytes_sent <= static_cast<int64_t>(data.size()));
+  if (stream_tx->bytes_sent + WireWriterImpl::kBlockSize >=
+      static_cast<int64_t>(data.size())) {
+    // This is the last transaction. Include trailing
+    // metadata if there's any.
+    if (tx->GetFlags() & kFlagSuffix) {
+      flags |= kFlagSuffix;
     }
-  } while (num_outgoing_bytes_ >=
-           num_acknowledged_bytes_ + kFlowControlWindowSize);
-  return true;
+    size = data.size() - stream_tx->bytes_sent;
+    *is_last_chunk = true;
+  } else {
+    // There are more messages to send.
+    flags |= kFlagMessageDataIsPartial;
+    *is_last_chunk = false;
+  }
+  RETURN_IF_ERROR(parcel->WriteInt32(flags));
+  RETURN_IF_ERROR(parcel->WriteInt32(next_seq_num_[tx->GetTxCode()]++));
+  if (flags & kFlagPrefix) {
+    RETURN_IF_ERROR(WriteInitialMetadata(*tx, parcel));
+  }
+  RETURN_IF_ERROR(parcel->WriteByteArrayWithLength(
+      data.substr(stream_tx->bytes_sent, size)));
+  if (flags & kFlagSuffix) {
+    RETURN_IF_ERROR(WriteTrailingMetadata(*tx, parcel));
+  }
+  stream_tx->bytes_sent += size;
+  return absl::OkStatus();
 }
 
-absl::Status WireWriterImpl::RpcCall(const Transaction& tx) {
-  // TODO(mingcl): check tx_code <= last call id
-  grpc_core::MutexLock lock(&mu_);
-  GPR_ASSERT(tx.GetTxCode() >= kFirstCallId);
-  if (CanBeSentInOneTransaction(tx)) {
-    return RpcCallFastPath(tx);
-  }
-  // Slow path: the message data is too large to fit in one transaction.
-  int& seq = seq_num_[tx.GetTxCode()];
-  int original_flags = tx.GetFlags();
-  GPR_ASSERT(original_flags & kFlagMessageData);
-  absl::string_view data = tx.GetMessageData();
-  size_t bytes_sent = 0;
-  while (bytes_sent < data.size()) {
-    if (!WaitForAcknowledgement()) {
-      return absl::InternalError("Timeout waiting for acknowledgement");
-    }
-    RETURN_IF_ERROR(binder_->PrepareTransaction());
-    WritableParcel* parcel = binder_->GetWritableParcel();
-    size_t size =
-        std::min(static_cast<size_t>(kBlockSize), data.size() - bytes_sent);
-    int flags = kFlagMessageData;
-    if (bytes_sent == 0) {
-      // This is the first transaction. Include initial metadata if there's any.
-      if (original_flags & kFlagPrefix) {
-        flags |= kFlagPrefix;
-      }
+void WireWriterImpl::RunScheduledTxInternal(RunScheduledTxArgs* args) {
+  GPR_ASSERT(args->writer == this);
+  if (absl::holds_alternative<RunScheduledTxArgs::AckTx>(args->tx)) {
+    int64_t num_bytes =
+        absl::get<RunScheduledTxArgs::AckTx>(args->tx).num_bytes;
+    absl::Status result = MakeBinderTransaction(
+        BinderTransportTxCode(BinderTransportTxCode::ACKNOWLEDGE_BYTES),
```

Comment:
This is already BinderTransportTxCode type.

Reply:
Done